### ISSUE TITLE
SourceMaps don't bind on latest next.js version server side (Docker)

### ISSUE THREAD

**Issue by jasonwilliams** [microsoft/vscode-js-debug#854]
**Describe the bug**
Next.js now uses `eval-source-map` from version 9 onwards.
It looks like `eval-source-map` generates `webpack-internal:///./` references.

I added the settings below but it makes no difference:
```json
      "sourceMapPathOverrides": {
        "webpack-internal:///./": "${workspaceFolder}/"
      }
```

I wonder if https://github.com/microsoft/vscode-js-debug/blob/e2191de3d65c8e90be183bd545ec8441972566a2/src/targets/browser/browserPathResolver.ts#L111-L113 has anything to do with it?

Should `webpack-internal:///./` be added to this list?
https://github.com/microsoft/vscode-js-debug/blob/4d7737fbb65cf6485bd333e03f5319ceafaa797e/src/configuration.ts#L931-L939

**To Reproduce**
1. Clone https://github.com/jasonwilliams/nextjs9-typescript-server-vscode-example
2. yarn install
3. `docker-composer up`
4. Launch -> launch server (Docker)
5. set a breakpoint in pages/index.tsx

**Log File**

**VS Code Version:** 1.51

**Comment by connor4312** [comment]
Thanks, this is actually an issue with `resolveSourceMapLocations` being narrowed. Previously for some reason I only removed the narrowing for `launch` rather than `attach` types. But I don't recall why I did this and I think it's wrong. Fixed in the next nightly.

**Comment by jasonwilliams** [comment]
This isn't fixed for me

![image](https://user-images.githubusercontent.com/936006/99064395-d47d3280-259d-11eb-9630-95b60de6ab96.png)

My nightly version is `2020.11.1117`

**Comment by connor4312** [comment]
To confirm, did you remove the `sourceMapPathOverrides` in the `launch server (Docker)` configuration?

**Comment by jasonwilliams** [comment]
@connor4312 I can confirm its the same whether I have `sourceMapPathOverrides` set or not.

**Comment by jasonwilliams** [comment]
@connor4312  This is still broken for me on `v2020.11.2317` can we re-open this issue?

**Comment by jasonwilliams** [comment]
I've had a go at debugging this, and dusted off my local copy of js-debug. It looks like `urlToAbsolutePath` fails with `webpack-internal://` urls. 

You can see here it returns an empty string:
![image](https://user-images.githubusercontent.com/936006/100471938-3330c900-30d3-11eb-9007-e9fab39c8497.png)

However for other files it works fine:
<img width="1347" alt="Screenshot 2020-11-27 at 17 10 07" src="https://user-images.githubusercontent.com/936006/100472227-bfdb8700-30d3-11eb-8c2f-1c8e3d92a5de.png">

**Comment by connor4312** [comment]
Thanks for the issue. Apologies for the delay, I've been on vacation.

Interestingly, it looks like next.js builds differently depending on whether I clone and run it (yes, in Docker) on Windows versus OSX. On Windows, `"file:///app/.next/server/static/development/pages/index.js"` has the sourceMap. On OSX, it uses an eval sourcemap where the sourceURL of the evaluated script is set to `webpack-internal:///./pages/index.tsx`, which is not mapped by the default source map resolution paths. Setting `resolveSourceMapLocations: null` resolves the issue.

This is messy to solve, since setting the sourceURL in this way entirely overwrites the original location we otherwise would see over CDP. The goal of `resolveSourceMapLocations` was to avoid scripts in dependencies that are packaged and include sourcemaps from incorrectly mapping to user code. It seems like we need to allow `webpack-internal:///*` to map by default, and this will break that safety net.

But there are more people using next.js than there are people who will hit that issue.

**Comment by connor4312** [comment]
Actually, the scenario is narrower than I thought. For data URI sourcemaps, we use the path of the compiled file to see if we should resolve it. I put in a fix to remove the "protocol", so `webpack-internal:///./pages/index.tsx` will turn into `/pages/index.tsx` for matching purposes, which the default glob is capable of dealing with. So I put in a targeted fix.

**Comment by jasonwilliams** [comment]
> Interestingly, it looks like next.js builds differently depending on whether I clone and run it (yes, in Docker) on Windows versus OSX. On Windows, "file:///app/.next/server/static/development/pages/index.js" has the sourceMap. On OSX, it uses an eval sourcemap where the sourceURL of the evaluated script is set to webpack-internal:///./pages/index.tsx, which is not mapped by the default source map resolution paths. Setting resolveSourceMapLocations: null resolves the issue.

Yep that sounds about right, they made that change intentionally here.
https://github.com/vercel/next.js/blob/7ce000b328305737099549bc186ddd3a39a2528b/packages/next/build/webpack/config/blocks/base.ts#L25-L36

**Comment by jasonwilliams** [comment]
Thanks @connor4312 appreciate you looking at this, and hope you had a good vacation!

**Comment by connor4312** [comment]
Windows is always fun 🙂 Thanks for the info and wishes, spent a week hiking on the Oregon coast.

**Comment by jasonwilliams** [comment]
> Setting resolveSourceMapLocations: null resolves the issue

Hey @connor4312 could you explain why setting this is required in NextJS projects? So hopefully i can add it to their documentation as i don't think this will be obvious to newcomers who want to debug

**Comment by roblourens** [comment]
[vscode-debugadapter-b67ca636.json.gz](https://github.com/microsoft/vscode-js-debug/files/5897174/vscode-debugadapter-b67ca636.json.gz)

I still see an unresolved breakpoint

**Comment by connor4312** [comment]
@roblourens make sure to remove the `sourceMapPathOverrides` section as well -- apologies for not posting easy verification steps in a long thread:

1. Follow these instructions: https://github.com/microsoft/vscode-js-debug/issues/854#issue-740838182
2. But before step 4, remove these lines https://github.com/jasonwilliams/nextjs9-typescript-server-vscode-example/blob/f404275fb07599b4a8e4135e10d2a8d0fc10163d/.vscode/launch.json#L22-L24

Closing for re-verification attempt

**Comment by usuallyno** [comment]
> This isn't fixed for me
> 
> ![image](https://user-images.githubusercontent.com/936006/99064395-d47d3280-259d-11eb-9630-95b60de6ab96.png)
> 
> My nightly version is `2020.11.1117`

I've exactly the same bug but with the client side launcher.
![screenshot](https://i.imgur.com/WmRtMLK.png)
It seems that "webpack://" is somehow mixed with the url..
How did you solve the problem?

my launch.json
```
   {
      "name": "Next.js: debug client-side",
      "type": "chrome",
      "request": "launch",
      "url": "http://localhost:3000",
      "skipFiles": [
        "<node_internals>/**",
        "${workspaceFolder}/node_modules/**/*.js",
        // "${workspaceFolder}/next"
      ]
    },
```

**Comment by ericdvb** [comment]
I'm not able to bind breakpoints for server side debugging in latest next.js (14.0.1 as of this writing, but also 13.5.6) using vs code. Jumping on this thread because it's unclear to me whether this is the same bug or something else.

Here's the relevant part of my `launch.json`:
```"version": "0.2.0",
    "configurations": [
        {
            "name": "Next.js: debug server-side",
            "type": "node",
            "address": "127.0.0.1",
            "port": 9229,
            "request": "attach",
            "sourceMaps": true,
            "skipFiles": ["<node_internals>/**"]
        },
...
```

my `tsconfig.json` includes `"sourceMap:" true`. I'm running my app using the following command: `"NODE_OPTIONS='--inspect=9229' next dev`.

I'm able to attach to the debugger, I see a log about it in the terminal window for my next app. If I use `--inspect-brk`, execution stops a next internals line (node_modules/next/dist/bin/next, I think). But any breakpoint I set in my api route (which is in `app/api/.../route.ts`) doesn't bind, and the troubleshooter shows some info about including `"sourceMap": true` in my tsconfig. Here's what it looks like in vs code in my file, and in the troubleshooter:
<img width="651" alt="Screen Shot 2023-11-01 at 8 32 00 AM" src="https://github.com/microsoft/vscode-js-debug/assets/4971172/cd1123f6-64c2-422a-bde4-286689b5b4bb">
<img width="647" alt="Screen Shot 2023-11-01 at 8 33 39 AM" src="https://github.com/microsoft/vscode-js-debug/assets/4971172/964b37a6-0ac8-4216-959e-2b017a107dec">
<img width="1103" alt="Screen Shot 2023-11-01 at 8 34 04 AM" src="https://github.com/microsoft/vscode-js-debug/assets/4971172/f1aa0dd3-6cf3-4ef7-8b7c-02a3fa6d4cfa">

I've scoured the next.js debug docs, a bunch of github issues, and plenty of people's posts about similar issues (although they're mostly from 2022, so it's unclear to me if they're relevant to newer versions of next or setups that use the app folder. I've tried adding `sourceMapPathOverrides` with the value `"webpack-internal:///./": "${workspaceFolder}/"` to my `launch.json`, and maybe a couple of other solutions, but no luck.

Is it possible to use breakpoint debugging with app folder api routes in next 14.0.1?
